lib/sqli/sqli_lexer.re2c

```diff
@@ -75,6 +75,16 @@ sqli_get_token(
 #define RET_DATA(toksuff, ctx, arg) \
     RET_DATA_EX(toksuff, ctx, arg, 0)
 #define RET_DATA_ERROR(ctx) detect_buf_deinit(&(ctx)->lexer.buf)
+#define DETECT_RE2C_YYFILL(lexer, pend, need, maxfill) \
+    do { \
+        int rv; \
+        if ((rv = detect_re2c_yyfill(&lexer.re2c, pend, need, maxfill))) { \
+            if (lexer.instring && lexer.re2c.fin) \
+                return TOK_DATA; \
+            else \
+                return (-rv); \
```

Comment:
YYFILL returning a token looks like a dirty hack.

Reply:
There is no other way to do the following logic:
1. Parse the injection `1 '^'`
2. When parsing starting from the string we have `'1' ^ '`
3. At the end of the parsing, we have a non-terminated string literal, which can be completed further. For example: `'1' ^ '2'`

Maybe you just need to change the name of the macro?
I will add tests
I will try to make more atomic changes

Reply:
Perhaps there is another way if we break the string literal into three lexemes (`'`, `content` and `'`)

Reply:
The tokens generation should occur in re2c rules. YYFILL is called by re2c but it is not a re2c rule. YYFILL actions are documented by re2c. YYFILL should just fill, it shouldn't return tokens.

The problem is the current alphabet doesn't contain 'nil' character. But it is rather easy to emulate it by checking 'fin' flag and the position of the current '\0' character. So, `<INSTRING,SQUOTE,DQUOTE> [\x00]` rule shouldn't be removed, it just have to be modified.

Reply:
Interesting. Thank you. I will try